src/client/failure/event-adapter.ts

```diff
@@ -17,24 +22,239 @@ CONSIDER DISABLING SENSITIVE INFORMATION LOGGING BY APPENDING THE DEBUG ENVIRONM
-      // tells post-perform to do provider recache and retry
-      | { kind: 'recache'; newRegistry?: string };
+    // action queued from nested hooks for the last hook
+    queuedAction: QueuedAction;
   }
 >;
 
+function handleCommonResolution(
+  performContext: HooksContext[string],
+  resolution: AbortResolution | RecacheResolution | SwitchProviderResolution
+): { newResult: Promise<never> } {
+  switch (resolution.kind) {
+    case 'abort':
+      performContext.queuedAction = {
+        kind: 'full-abort',
+        reason: resolution.reason,
+      };
+
+      return {
+        // this error will be overridden in post-bind-and-perform
+        newResult: Promise.reject('full abort in progress'),
+      };
+
+    case 'recache':
+      performContext.queuedAction = resolution;
+
+      return {
+        // this error will be overridden in post-bind-and-perform
+        newResult: Promise.reject('recache in progress'),
+      };
+
+    case 'switch-provider':
+      performContext.queuedAction = resolution;
+
+      return {
+        // this error will be overridden in post-bind-and-perform
+        newResult: Promise.reject('failover in progress'),
+      };
+  }
+}
+
 export function registerHooks(hookContext: HooksContext, events: Events): void {
+  registerNetworkHooks(hookContext, events);
+
+  events.on('pre-bind-and-perform', { priority: 1 }, async (context, args) => {
+    debug('Handling event pre-bind-and-perform with context:', context);
+    debugSensitive('\targs:', args);
+
+    if (context.profile === undefined || context.usecase === undefined) {
+      throw new UnexpectedError('invalid event context');
+    }
+
+    // only check failover restore when the provider is not manually set
+    if (args[1]?.provider !== undefined) {
+      return { kind: 'continue' };
+    }
+
+    const performContext = hookContext[`${context.profile}/${context.usecase}`];
+    // if there is no configured context: ignore the event
+    if (performContext === undefined) {
+      return { kind: 'continue' };
+    }
+
+    if (performContext.router.getCurrentProvider() === undefined) {
+      return { kind: 'continue' };
+    }
+
+    const resolution = performContext.router.beforeExecution({
+      time: context.time.getTime(),
+      registryCacheAge: 0, // TODO
+      checkFailoverRestore: true,
+    });
+
+    switch (resolution.kind) {
+      case 'continue':
+        return { kind: 'continue' };
+
+      case 'backoff':
+        return { kind: 'continue' };
+
+      default:
+        return {
+          kind: 'abort',
+          ...handleCommonResolution(performContext, resolution),
+        };
+    }
+  });
+
+  events.on(
+    'post-bind-and-perform',
+    { priority: 1 },
+    async (context, args, res) => {
+      debug('Handling event post-bind-and-perform with context:', context);
+      debugSensitive('\targs:', args);
+      debugSensitive('\tresult:', res);
+
+      if (
+        context.profile === undefined ||
```

Comment:
Huh, so the problem goes a bit deeper.

The `pre-bind-and-perform` hook doesn't know `provider`. The `post-bind-and-perform` hook knows provider only if the `pre-` hook didn't abort (because the `provider` field is filled in the bind portion of the `bindAndPerform` method).

The `post-` hook needs to know the provider for metrics basically.

There is a possiblity of a broken `pre-` hook aborting the wrong way and thus leaving `context.provider` empty for the `post-` hook to find. So I made `profile` and `usecase` context values required but I think we have to leave the `provider` value optional.

Reply:
ok